blueprints/snapshots.and.monitoring.md

```diff
@@ -0,0 +1,828 @@
+			"localizationMethods": [
+				"GEO",
+				"CZ",
+				"DEEP_CZ"
+			]
+		}
+	},
+	"trafficRouterCacheGroups": {
+		"CDN_in_a_Box_Edge": {
+			"latitude": 38.897663,
+			"longitude": -77.036574,
+			"trafficRouters": {
+				"trafficrouter": {
+					"apiPort": 3333,
+					"fqdn": "trafficrouter.infra.ciab.test",
+					"httpsPort": 443,
+					"ipv4Address": "172.19.0.14",
+					"ipv6Address": null,
+					"secure.api.port": 3443,
+					"tcpPort": 80
+				}
+			}
+		}
+	},
+	"trafficMonitors": {
+		"trafficmonitor": {
+			"fqdn": "trafficmonitor.infra.ciab.test",
+			"httpsPort": 443,
+			"ipv4Address": "172.19.0.12",
+			"ipv6Address": null,
+			"status": "ONLINE",
+			"tcpPort": 80
+		}
+	},
+	"stats": {
+		"cdn": "CDN-in-a-Box",
+		"date": 1607615441,
+		"toHost": "trafficops.infra.ciab.test:443"
+	},
+	"topologies": {
+		"demo1-top": {
+			"nodes": [
+				"CDN_in_a_Box_Edge"
+			]
+		}
+	}
+}}
+```
+
+## Traffic Portal Impact
+Traffic Portal shouldn't _need_ any changes, but it should ideally show
+Monitoring payloads in its Snapshot diff page, and there's never a better time
+than when big changes are being made to both, to help reduce operaton
+confusion.
+
+## Traffic Ops Impact
+### API Impact
+The CDN Snapshot and Monitoring handlers will need to be updated to output the
+new formats, which will necessitate a new major API version. This includes the
+following endpoints:
+
+- `/cdns/{{name}}/configs/monitoring`
+- `/cdns/{{name}}/snapshot/new`
+- `/cdns/{{name}}/snapshot`
+- `/snapshot` (this, however, shares most of its logic with `/cdns/{{name}}/snapshot/new`)
+
+New logic will also need to be added to convert from the new snapshot (and
+monitoring payload) formats back to the old formats, so that newly stored
+Snapshot (and monitoring payload) data can be served in old API versions.
+
+### Client Impact
+The APIv4 client will need to change the call signatures of the affected
+endpoint methods.
+
+### Database Impact
+Existing stored Snapshots (and monitoring payload data) will need to be
+converted to the new format via a migration. No schema changes will be
+necessary, because Snapshots (and monitoring payload data) are stored as
+JSON blobs.
+
+## ORT Impact
+None.
+
+## Traffic Monitor Impact
+Traffic Monitor will need to be overhauled; Snapshot dependency is ingrained
+fairly deep in it. It will need to build out its configuration and stats from
+only the monitoring payload. It will also need to keep serving the Snapshots,
+however, to support old versions of Traffic Router.
+
+Traffic Monitor's parsing of monitoring payloads will also need to be updated
+to handle the new format - but as always, a previous API version must still be
+supported.
+
+## Traffic Router Impact
+Traffic Router will need to be updated to handle the new Snapshot format, as
+well as maintaining compatibility with the old format. It will also need to be
+updated to fetch Snapshots from Traffic Ops directly, rather than through
+Traffic Monitor.
```

Comment:
TR has the nice ability today to be able to retrieve a snapshot from any of the Traffic Monitors. How does replacing n TMs with 1 TO impact the resiliency of config update process?

Reply:
In terms of resiliency,
If TO fails, the TMs won't be able to get config either. So it's not really introducing a single point of failure for new config. That already exists.

But what it does do is introduce a point-of-failure which can make the config race condition permanent. That is, if TM gets a new config, and TO goes down before TR can get the same config (or vica versa), what was a temporary race can become a long-standing mismatch, potentially outage-inducing, until TO is back up.

Where with the TO->TM->TR flow, no race exists, so TR will always match TM's config. If TO goes down, it may not be the latest, but it won't be an outage-inducing mismatch.

Reply:
Per my comment on the main thread, I think TM and TR can safely get out of sync without causing an outage.

Reply:
I'm just going to resolve this in favor of the longer discussion below